test/unit/specs/core/createEventManager.spec.js

```diff
@@ -108,31 +108,51 @@ describe("createEventManager", () => {
         });
     });
 
-    it("sets the lastChanceCallback, which wraps config.onBeforeEventSend, on the event", () => {
-      let wrappedLastChanceCallback;
-      event.setLastChanceCallback.and.callFake(callback => {
-        wrappedLastChanceCallback = callback;
+    it("events call onBeforeEventSend callback", () => {
+      event.finalize.and.callFake(onBeforeEventSend => {
+        onBeforeEventSend();
       });
-      return eventManager.sendEvent(event, {}).then(() => {
-        wrappedLastChanceCallback();
+      return eventManager.sendEvent(event).then(() => {
         expect(config.onBeforeEventSend).toHaveBeenCalled();
       });
     });
 
-    it("logs errors in the config.onBeforeEventSend callback", () => {
-      const error = new Error("onBeforeEventSend error");
-      config.onBeforeEventSend.and.throwError(error);
+    it("returns false onBeforeEventSend and event should not be sent", () => {
+      const eventToIgnore = jasmine.createSpyObj("event", {
+        finalize: () => {},
+        shouldSend: false
+      });
+      return eventManager.sendEvent(eventToIgnore).then(() => {
+        expect(sendEdgeNetworkRequest).not.toHaveBeenCalled();
+      });
+    });
 
-      let wrappedLastChanceCallback;
-      event.setLastChanceCallback.and.callFake(callback => {
-        wrappedLastChanceCallback = callback;
+    it("returns null onBeforeEventSend and event should be sent", () => {
+      const eventToIgnore = jasmine.createSpyObj("event", {
+        finalize: () => {},
+        shouldSend: null
+      });
+      return eventManager.sendEvent(eventToIgnore).then(() => {
+        expect(sendEdgeNetworkRequest).toHaveBeenCalled();
       });
+    });
+
+    it("returns true onBeforeEventSend and event should be sent", () => {
+      const eventToIgnore = jasmine.createSpyObj("event", {
+        finalize: () => {},
+        shouldSend: true
+      });
+      return eventManager.sendEvent(eventToIgnore).then(() => {
+        expect(sendEdgeNetworkRequest).toHaveBeenCalled();
+      });
+    });
+
+    it("throws an error on event finalize and event should not be sent", () => {
+      const errorMsg = "Expected Error";
+      event.finalize.and.throwError(errorMsg);
 
-      return eventManager.sendEvent(event, {}).then(() => {
-        expect(() => {
-          wrappedLastChanceCallback();
-        }).toThrowError("onBeforeEventSend error");
-        expect(logger.error).toHaveBeenCalledWith(error);
+      return eventManager.sendEvent(event).catch(() => {
+        expect(sendEdgeNetworkRequest).not.toHaveBeenCalled();
```

Comment:
You'll need to make sure that the promise returned to Jasmine is rejected if the promise from `eventManager.sendEvent(event)` is resolved. Here's why: Jasmine sees you returned a promise and will wait until the promise is resolved or rejected before moving to the next test. If the promise is resolved and all assertions pass, then it considers the test as passed. If the promise is rejected or any assertion fails, then it considers the test as failed. The problem here is that if the promise coming from `eventManager.sendEvent(event)` is resolved rather than rejected, then your `.catch` will never be called and the promise jasmine is observing will be resolved. Therefore, the test would be considered as passed, when it should be considered as failed. I would recommend something like this:

```js
return eventManager.sendEvent(event).then(
  () =>
    throw new Error("Should not have resolved.")
  }, 
  () => {
    expect(sendEdgeNetworkRequest).not.toHaveBeenCalled();
  }
);

Reply:
I think another options is using `expectAsync`. It works something like this:

await expectAsync(boom()).toBeRejectedWith(new TypeError("BOOM"))

and then you could do `expect(sendEdgeNetworkRequest).not.toHaveBeenCalled();` afterward.